### PR TITLE
Add RequestStarted to GlobalStatusCode enum

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
This PR adds support for moving the initial "request supported" handshake into the swarm. The request should be adapted so they don't send the request implementation defined `Started` code, but to just send `RequestStarted` when they are ready to start handling the request. Then they can send `Error` message (which is request defined) any time to the client, which should also expect it and abort everything with the `node_error`.

This is to allow request handlers to be `nothrow`, notifying the client about the error that just happened, and which can happen any time.

Node's connection handler then can be enhanced with the nothrow enforcement, but that needs to be done in the next major.

### CODE REVIEW

**Review by gavin-norman-sociomantic — COMMENTED** [review]
A high-level comment that applies to all commits:

I'd suggest that we rename things a bit:
* `GlobalStatusCode` -> `SupportedCode`.
* `RequestStarted` -> `RequestSupported`.
* `handleRequestStatusCodes` -> `handleSupportedCodes`

The current sole function of this code is to indicate supported / not, so unless there are any ideas for it being used to carry any other messages, we may as well make it clear what it represents.

**Comment by nemanja-boric-sociomantic** [comment]
Updated and rebased to v4.x.x. I'm not sure what to do about `GlobalStatusCode` enum. I feel `deprecating` it here would bring more harm than good, as it would propagate to all current requests.

**Comment by nemanja-boric-sociomantic** [comment]
(and it's not an internal request detail, like `nextEvent`, but actually a part of the node/client protocol.)

File: src/swarm/neo/request/Command.d [hunk 1]
```diff
@@ -61,6 +61,11 @@ public alias byte StatusCode;
 
 public enum GlobalStatusCode : StatusCode
 {
+    /// Both request code specified in the transmitted Command struct and the
+    /// associated request version are supported and the request will start
+    /// on this connection.
+    RequestSupported = -3,
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/request/Command.d:77]: The commit message of "Add RequestStarted to GlobalStatusCode enum" is outdated.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 2]
```diff
@@ -375,6 +375,57 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
...
+            remote_address = address of node which returned this status code
+
+        Returns:
+            false if `status` was a global error code and the user notified;
+            true if `status` is a RequestSupported code indicating that the
+            request is supported and that the request will start.
+
+    ***************************************************************************/
+
+    private static bool handleSupportedCodes ( StatusCode status,
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: As we no longer intend to allow the set of codes to be "extended" per request type, this function should just accept `SupportedStatus` directly.

File: src/swarm/neo/request/Command.d [hunk 3]
```diff
@@ -61,6 +61,11 @@ public alias byte StatusCode;
 
 public enum GlobalStatusCode : StatusCode
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/request/Command.d]: Maybe add `alias GlobalStatusCode SupportedStatus;`.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 4]
```diff
@@ -375,6 +375,57 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
...
+        Private helper function to handle notification of the global supported
+        codes, including the RequestSupported code.
+
+        Params:
+            status = code received from node
+            context = deserialized request context
+            remote_address = address of node which returned this status code
+
+        Returns:
+            false if `status` was a global error code and the user notified;
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: "global error code" is outdated terminology.

File: src/swarm/neo/request/Command.d (hunk 3 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/request/Command.d]: Oh, or would renaming the enum and adding a deprecated alias do the job? (As long as we change all usages inside swarm, we shouldn't get deprecation warnings.)

File: integrationtest/neo/client/request/internal/GetAll.d [hunk 5]
```diff
@@ -275,32 +275,11 @@ private scope class GetAllImpl
     {
         auto getall_status = cast(RequestStatusCode)status;
```
Comment (gavin-norman-sociomantic) [inline integrationtest/neo/client/request/internal/GetAll.d]: This variable is a bit deceptive now, as there are no special GetAll status codes.

File: integrationtest/neo/client/request/internal/GetAll.d [hunk 6]
```diff
@@ -393,21 +373,34 @@ private struct Reader
...
-        // Acknowledge the End signal. The protocol guarantees that the node
-        // will not send any further messages.
-        this.request_event_dispatcher.send(this.fiber,
-            ( RequestOnConn.EventDispatcher.Payload payload )
-            {
-                payload.addConstant(MessageType.Ack);
-            }
-        );
-        this.conn.flush();
+        if (!error)
```
Comment (gavin-norman-sociomantic) [inline integrationtest/neo/client/request/internal/GetAll.d]: Could this be moved inside `case End:`?

File: integrationtest/neo/node/request/GetAll.d [hunk 7]
```diff
@@ -38,15 +38,23 @@ public void handle ( Object shared_resources, RequestOnConn connection,
...
+
     switch ( cmdver )
     {
         case 0:
+            ed.send(
+                ( ed.Payload payload )
+                {
+                    payload.addConstant(GlobalStatusCode.RequestSupported);
+                }
+            );
```
Comment (gavin-norman-sociomantic) [inline integrationtest/neo/node/request/GetAll.d]: This ought to be flushed.

Comment (gavin-norman-sociomantic) [reply integrationtest/neo/node/request/GetAll.d]: (All the more reason for https://github.com/sociomantic-tsunami/swarm/issues/168.)

File: integrationtest/neo/node/request/GetAll.d [hunk 8]
```diff
@@ -218,33 +226,36 @@ private scope class GetAllImpl_v0
...
+                this.writer.suspender.requestSuspension();
 
-        this.controller.fiber.start();
-        this.writer.fiber.start();
-        this.request_event_dispatcher.eventLoop(this.conn);
+            this.controller.fiber.start();
+            this.writer.fiber.start();
+            this.request_event_dispatcher.eventLoop(this.conn);
+        }
+        catch (Exception e)
```
Comment (gavin-norman-sociomantic) [inline integrationtest/neo/node/request/GetAll.d:253]: I really wonder if this would after all be worth handling at a higher level. The only reason I can think of to not do that is that the exact code to send is defined per request (atm).

File: src/swarm/neo/request/Command.d (hunk 3 above)
Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/Command.d]: No, sadly, this is used outside the swarm (the node handlers are sending these codes to the clients).

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/request/Command.d]: Right, that would mean that when we update *proto to use swarm v4.5.0, we'd get deprecation warnings that we'd then fix. Or?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/request/Command.d]: Right, that would work, it's just renaming in this case.

File: integrationtest/neo/node/request/GetAll.d (hunk 8 above)
Comment (nemanja-boric-sociomantic) [reply integrationtest/neo/node/request/GetAll.d:253]: It's a tricky situation. The reason I didn't want it there, is that perhaps different requests may define different ways how the errors are reported to the client (and it requires change of the request definitions to expect it anytime). I guess I could add the separate method that runs the handler, accepts the error code (or use the swarm defined, in which case we'll have to add one), but then it's the question how should we plug this with the existing request.

Perhaps we should just add the request code and send it, as, if the client is not aware of it, it will shut down with the protocol error, which is still better than the current situation?

Comment (gavin-norman-sociomantic) [reply integrationtest/neo/node/request/GetAll.d:253]: I suppose we can leave it down to each request for now, then see what common pattern develops. (I suspect all requests will end up doing the same, but we may be surprised.)

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

File: integrationtest/neo/client/request/internal/GetAll.d [hunk 9]
```diff
@@ -273,7 +274,7 @@ private scope class GetAllImpl
 
     private bool handleStatusCode ( ubyte status )
     {
-        auto request_status = cast(RequestStatusCode)status;
+        auto request_status = cast(SupportedStatus)status;
```
Comment (gavin-norman-sociomantic) [inline integrationtest/neo/client/request/internal/GetAll.d]: This looks unrelated to this commit?

Comment (nemanja-boric-sociomantic) [reply integrationtest/neo/client/request/internal/GetAll.d]: Hm, I wanted to get rid of all the deprecations, but I should actually get rid of them in a good way - actually using the new mechanism.

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

**Comment by gavin-norman-sociomantic** [comment]
Hmm release notes? The deprecation could be mentioned, say.

**Comment by nemanja-boric-sociomantic** [comment]
Oh, damn, that slipped me!

**Comment by gavin-norman-sociomantic** [comment]
Yeah, and new feature.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with :police_car:

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 10]
```diff
@@ -375,6 +375,57 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
...
+            remote_address = address of node which returned this status code
+
+        Returns:
+            false if `status` was a status error code and the user notified;
+            true if `status` is a RequestSupported code indicating that the
+            request is supported and that the request will start.
+
+    ***************************************************************************/
+
+    private static bool handleSupportedCodes ( SupportedStatus status,
```
Comment (gavin-norman-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d:395]: Actually, is there any reason to not deprecate `handleGlobalStatusCodes`?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: Yes, we don't have suitable replacement for the old requests which can't yet move to the new handshake mechanism.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: Mmm, isn't `handleSupportedCodes` it?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: No, because `handleSupportedCodes` can't tell the client if the actual status code was (previously) `GlobalStatusCode` or a request specific code - it returns false for all of them (it returns true only if the `RequestStarted` happens, which is never the case in the old request). In contrast, `handleGlobalStatusCodes` would return `true` if the status code was a global status code, but `false` if it was the request-specific status code.

Perhaps the deprecation path could be checking if the code is positive after casting it to `SupportedStatus` (signed byte), but that doesn't strike me as a nice solution.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: Maybe I'm missing something...

A replacement for a deprecated method doesn't have to be a _direct_ replacement. It may require changes other than changing the name of the symbol used. The deprecation just indicates that something is changing and that all users of the code will have to migrate. Which is exactly the case we have here. Or?

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: The problem is that I don't see a way to fix the old, but yet supported requests, without the change in the _protocol_ (this is in contrast to our usual deprecations that can be fixed without changing the on-wire protocol). That means that either we need to make a new version of the request, or we're stuck with the silly deprecation warning for which the proto users (so, not swarm users, but the users of the concrete proto implementations) can't do anything. Now, making a new version of the request just for this change, to get rid of this particular deprecation might be fine, but I was erring on the side of not having to do this.

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: (I probably didn't use `erring` in a good context here).

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:395]: Ah I see. You're right.

**Review by gavin-norman-sociomantic — APPROVED** [review]
